update/cleanup dependencies

tchaikov opened the pull request.

kshtsk commented on requirements.txt:
```diff
@@ -38,8 +37,8 @@ first==2.0.1              # via pip-tools
 futures==3.2.0            # via s3transfer
-gevent==1.2.1
-greenlet==0.4.12          # via gevent
+gevent==1.4.0
```
These two good.

kshtsk commented:
@susebot run deploy

kshtsk commented on requirements.txt:
```diff
@@ -38,8 +37,8 @@ first==2.0.1              # via pip-tools
 funcsigs==1.0.2           # via debtcollector, oslo.utils, pytest
 functools32==3.2.3.post2  # via jsonschema
 futures==3.2.0            # via s3transfer
```
I just wonder if we can get rid of functools32 and futures here?

tchaikov replied:
will check tomorrow.

tchaikov replied:
no, we cannot. they are dependencies introduced by boto3 and python-openstackclient respectively.

kshtsk replied:
too bad, because those two keep me out of running tests in py3

tchaikov replied:
what do you mean by "running tests"? are these tests teuthology's own tests?

kshtsk replied:
I mean:
```
virtualenv -p python3.7
pip install -r requirements.txt 
```
fails for me

toabctl replied:
we need to tell pip that functools and futures are python2 only requrirements.

toabctl replied:
The problem is `pip-compile` itself. When you call it from a python3 env, you get:
```
(venv3) tom@minze:~/devel/ceph/teuthology$ pip-compile -P boto3 2>&1 | functools
```

and from a py2 venv:
```
(venv2) tom@minze:~/devel/ceph/teuthology$ pip-compile -P boto3 2>&1 | grep functools
functools32==3.2.3.post2  # via jsonschema
```

So imo `pip-compile` is broken by design (or there is some flag to fix this...)

tchaikov replied:
@kshtsk i feel sorry about your pain. but i think the long term solution is to ditch the python2 support.

kshtsk commented on setup.py:
```diff
@@ -53,47 +53,47 @@
                       ],
     extras_require = {
         'coverage': [ 'mysqlclient == 1.4.2'],
+        'orchestra': [
```
I guess documentation should be updated to reflect this change.

kshtsk replied:
If you run `pip-compile -P something` the orchestra deps are not be included in the requirements.txt?

tchaikov replied:
yes, the orchestra deps *should* be included. please use `mysqlclient` as an example, it's in "coverage", but `mysqlclient` is still included by `requirements.txt`.

tchaikov replied:
will update the doc.

tchaikov replied:
done.

tchaikov commented:
> - I don't see any import from `boto`, `boto3` or `xmltodict`.

@epuertat boto is used by some rgw-admin related test. boto3 and xmltodict are supposed to used by rgw pub/subscribe endpoint related tests in ceph/ceph repo. guess @yuvalif didn't get a chance to add it. or the tests have not been merged yet.

tchaikov commented:
@kshtsk @epuertat i 

- added another extra requirement for testing teuthology itself.
- updated `docs/INSTALL.rst` to document the extra requirements.

yuvalif commented:
> > have you measured how long does it take now to install it
> 
> not yet.
> 
> > * I don't see any import from `boto`, `boto3` or `xmltodict`.
> 
> @epuertat boto is used by some rgw-admin related test. boto3 and xmltodict are supposed to used by rgw pub/subscribe endpoint related tests in ceph/ceph repo. guess @yuvalif didn't get a chance to add it. or the tests have not been merged yet.

they are used in the multisite and pubsub tests in: src/test/rgw/rgw_multi/

kshtsk commented:
@tchaikov could you please add the command to the doc how to produce new requirement.txt with pip commands so it includes all those new sections?

Thanks.

kshtsk commented on setup.py:
```diff
@@ -93,6 +90,11 @@
+        'test': [
+            'tox',
+            'pip-tools',
+            'pytest',           # for tox.ini
```
Does it worth to add mock==2.0.0, fudge, pytest-cov==1.6, coverage==3.7.1 and something other from the .tox.ini here too?

tchaikov replied:
not sure which is better -- to document the dependencies in `setup.py` or in `tox.ini`. IMHO, the ones specified by `setup.py` are install dependencies which are not necessarily the dependencies for running tests, which are used by developers only. and requirements.txt is the right place for them. but in our case, requirements.txt is generated. so, `tox.ini` is good enough. i think.

kshtsk replied:
If so, why do you include pytest in setup.py?
From other side, I usually run pytest without tox when fixing them and debugging, so I have to manually install mock, flake8 and the rest.

tchaikov replied:
@kshtsk it was included by "install_requires". please see my commit message

> setup.py: move test related stuff to optional deps

i intended to move them instead of removing them.

tchaikov commented:
> @tchaikov could you please add the command to the doc how to produce new requirement.txt with pip commands so it includes all those new sections?
> 
> Thanks.

i think `requirement.txt` is self-documented, please see the header of this file. and all sections are always included even if they are not specified.

tchaikov commented on docs/INSTALL.rst:
```diff
@@ -54,3 +54,9 @@ However if you prefer, you may install ``teuthology`` from `PyPI <http://pypi.py
 
+Or from GitHub::
+
+    pip install git+https://github.com/ceph/teuthology#egg=teuthlogy[orchestra]
```
fixed and repushed.

kshtsk commented:
> > @tchaikov could you please add the command to the doc how to produce new requirement.txt with pip commands so it includes all those new sections?
> > Thanks.
> 
> i think `requirement.txt` is self-documented, please see the header of this file. and all sections are always included even if they are not specified.

well, I've run `pip-compile` from your branch and it produced a file without all dependecies you've moved away to separate extras, what I did wrong?

kshtsk commented:
@tchaikov which version of pip-compile do you use to generate the requirements.txt?

When I run bootstrap, and try to run pip-compile I get this:
```
$ pip-compile --version
pip-compile, version 4.3.0

$ pip-compile
#
# This file is autogenerated by pip-compile
# To update, run:
#
#    pip-compile
#
ansible==2.8.2
argparse==1.4.0
asn1crypto==0.22.0        # via cryptography
boto3==1.9.161
boto==2.46.1
botocore==1.12.161        # via boto3, s3transfer
cffi==1.10.0              # via cryptography
configobj==5.0.6
configparser==3.5.0
cryptography==2.7         # via ansible
docopt==0.6.2
docutils==0.14            # via botocore
enum34==1.1.6             # via cryptography
futures==3.2.0            # via s3transfer
gevent==1.4.0
greenlet==0.4.15          # via gevent
ipaddress==1.0.18         # via cryptography
jinja2==2.10.1            # via ansible
jmespath==0.9.4           # via boto3, botocore
manhole==1.3.0
markupsafe==1.0           # via jinja2
nose==1.3.7
pexpect==4.2.1
prettytable==0.7.2
psutil==5.2.2
ptyprocess==0.5.1         # via pexpect
pycparser==2.17           # via cffi
python-dateutil==2.6.0    # via botocore
pyyaml==5.1.2
s3transfer==0.2.1         # via boto3
six==1.10.0
urllib3==1.25.3           # via botocore
xmltodict==0.12.0

$ git diff --stat
 requirements.txt | 81 +++++----------------------------------------------------------------------------
 1 file changed, 5 insertions(+), 76 deletions(-)

```

kshtsk commented:
@tchaikov ping, I still don't know how you were able to generate the requirements.txt? Any thoughts?

tchaikov commented:
@kshtsk sorry for the latency. will get back to this PR tomorrow.

tchaikov commented:
@kshtsk sorry, i treated.
```
$ cat requirements.in
-e .[coverage,orchestra,test]
```

kshtsk commented:
> @kshtsk sorry, i cheated.
> 
> ```
> $ cat requirements.in
> -e .[coverage,orchestra,test]
> ```

are you going to add the `requirements.in` within this PR or update the docs?

kshtsk commented on docs/INSTALL.rst:
```diff
@@ -60,3 +60,13 @@ Or from GitHub::
+
+We track the dependencies using ``requirements.txt``. These packages are tested,
+and should work with teuthology. But if you want to bump up the version of them,
+please use follow command to update this file ::
```
follow => the following

tchaikov replied:
fixed and repushed.

tchaikov commented:
@kshtsk instead i updated the doc. because, if we include `requirements.in` and use `pip-compile` directly, a

```
-e file:///path/to/your/teuthlogy
```

line will show up in the generated `requirements.txt`.

kshtsk commented:
> @kshtsk instead i updated the doc. because, if we include `requirements.in` and use `pip-compile` directly, a
> 
> ```
> -e file:///path/to/your/teuthlogy
> ```
> 
> line will show up in the generated `requirements.txt`.

Is this too bad?

tchaikov commented:
@kshtsk yes, too bad. every time, a teuthology developer is required to edit the generated file manually. and the reviewer might need to remind the author of a PR to remove the line.`

kshtsk commented on docs/INSTALL.rst:
```diff
@@ -60,3 +60,13 @@ Or from GitHub::
+and should work with teuthology. But if you want to bump up the version of them,
+please use the following command to update this file ::
+
+  echo "-e .[coverage,orchestra,test]" | pip-compile - -qo- | sed '/^-e / d' > requirements.txt
```
What about changing this to:
```
pip-compile -qo- | sed '/^-e / d' > requirements.txt
```
and commit `requirements.in`:
```
-e .[coverage,orchestra,test]
```
Also we can have `update_requirements` shell script in the root directory?

tchaikov replied:
i think the document is good enough. if we really believe the usability the one-liner shell script is a pain, i will fix it in a follow-up PR.

kshtsk replied:
I'm fine with the oneliner, trying to find better to way to manage `coverage,orchestra,test` dependencies. It is more intuitive to look for them in requirements.in instead of modifying the doc each time new dependency is added or removed.

kshtsk replied:
Usually, updating the docs is the last thing a developer thinks.

tchaikov replied:
normally, it's not necessary to update the document for adding or removing dependencies. in this change, it's needed only if a new subset of extra requirement is added.

tchaikov replied:
normally, it's not necessary to update the document for adding or removing dependencies. in this change, it's needed only if a new *subset* of extra requirement is added.

tchaikov replied:
as you insist, i added `requirements.in` in the latest change.  but please note, it's only a workaround and a hack. and it's not quite right. because the `requirements.in` is pointing to the package itself. i'd avoid adding such a file if i could.

also, i don't want to add the helper script. and i explained the reason in https://github.com/ceph/teuthology/pull/1367#discussion_r353896681 . as the document already put the way how `requirements.txt` is updated. another script just adds another layer of abstraction which is overkill.

tchaikov commented:
kshtsk <notifications@github.com>于2019年12月5日 周四02:24写道：

> *@kshtsk* commented on this pull request.
> ------------------------------
>
> In docs/INSTALL.rst
> <https://github.com/ceph/teuthology/pull/1367#discussion_r353907255>:
>
> > @@ -60,3 +60,13 @@ Or from GitHub::
>
>  where the dependencies for orchestrating are installed. They are used for
>  interacting with the services to schedule tests and to report the test results.
> +
> +
> +Update Dependencies
> +-------------------
> +
> +We track the dependencies using ``requirements.txt``. These packages are tested,
> +and should work with teuthology. But if you want to bump up the version of them,
> +please use the following command to update this file ::
> +
> +  echo "-e .[coverage,orchestra,test]" | pip-compile - -qo- | sed '/^-e / d' > requirements.txt
>
> Usually, updating the docs is the last thing a developer thinks.
>

I am not sure what you mean here. Probably I don’t care what usually a
developer thinks. Also, please note I was following your suggestion in
another comment in this PR to update the document.

—
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ceph/teuthology/pull/1367?email_source=notifications&email_token=AAAONP4RCXNS4BQNQVA7JO3QW7YUZA5CNFSM4JREAV32YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCN7H37Y#discussion_r353907255>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAONP45WAOUEFLLSPKCYKLQW7YUZANCNFSM4JREAV3Q>
> .
>
-- 
Regards
Kefu Chai

kshtsk commented:
>>interacting with the services to schedule tests and to report the test results.
>>
>>```
>>+ + +Update Dependencies 
>>+------------------- 
>>+ 
>>+We track the dependencies using ``requirements.txt``.  These packages are tested, 
>>+and should work with teuthology. But if you want to bump up the version of them, 
>>+please use the following command to update this file :: 
>>+ 
>>+ echo "-e .[coverage,orchestra,test]" | pip-compile - -qo- | sed '/^-e / d' > requirements.txt 
>>```
>> Usually, updating the docs is the last thing a developer thinks.
>
> I am not sure what you mean here. Probably I don’t care what usually a developer thinks. Also, please note I was following your suggestion in another comment in this PR to update the document.

And great thanks for following, I'm just referring your comment:
> @kshtsk yes, too bad. every time, a teuthology developer is required to edit the generated file manually. and the reviewer might need to remind the author of a PR to remove the line.`

Anyways, thank for adding it to docs.

toabctl commented:
@tchaikov can't we just get rid of `pip-compile` and the `requirements.txt` file and only declare dependencies in `setup.py` ? That would simplify things (I know from time to time, pypi module updates will break things but that is then fixable quickly) and also would solve the py2/py3 problem where `pip-compile` is just broken...

tchaikov commented:
hi @toabctl , the original intention of this PR is to remove another hurdle of running teuthology based tests locally using python3. and my solution was to have a minimal set of dependencies in `install_requires`. but it seems the purpose of PR changes overtime. now it starts to cleanup `requirements.txt` and `setup.py` =)

> That would simplify things (I know from time to time, pypi module updates will break things but that is then fixable quickly) and also would solve the py2/py3 problem where pip-compile is just broken...

yeah, as we know the intention of having `requirements.txt` is to allow us to have a "snapshot" of known-to-work dependencies. so it's not in the scope of this PR, and i'd leave it to another developer. BTW, probably it's fixable quickly. but i am afraid that not all of us want to stop and start debugging teuthology if he/she in in the middle of something else, and just wants to schedule a test.

kshtsk commented:
> @tchaikov can't we just get rid of `pip-compile` and the `requirements.txt` file and only declare dependencies in `setup.py` ? That would simplify things (I know from time to time, pypi module updates will break things but that is then fixable quickly) and also would solve the py2/py3 problem where `pip-compile` is just broken...

this is a good idea, I was also thinking of it, maybe before I merge the rest of py3 stuff I will remove the requirements files totally, and we do not need to worry much about the requirements.txt or requirements.in too much

kshtsk commented:
@susebot run deploy

epuertat commented on docs/INSTALL.rst:
```diff
@@ -54,3 +54,19 @@ However if you prefer, you may install ``teuthology`` from `PyPI <http://pypi.py
+and should work with teuthology. But if you want to bump up the versions of them,
+please use the following command to update this file ::
+
+  pip-compile -qo- | sed '/^-e / d' > requirements.txt
```
Instead of this recipe and creating requirements.in, what about providing a wrapper (as supported by `pip-compile`) which also embeds extras?:
```sh
# file: update-requirements.sh, allows for 2 input env vars: $PYTHON_VERSION and $EXTRAS
PYTHON_VERSION=${PYTHON_VERSION:-3}
EXTRAS=${EXTRAS:-coverage,orchestra,test}
$ CUSTOM_COMPILE_COMMAND="PYTHON_VERSION=${PYTHON_VERSION} EXTRAS=${EXTRAS} generate-requirements.sh" python${PYTHON_VERSION} -mpiptools compile -qo- <(echo "-e .[${EXTRAS}]") | sed '/^-e / d' > requirements-py${PYTHON_VERSION}.txt
``` 

It would generate a requirements-py{27,3}.txt like this:
```sh
# This file is autogenerated by pip-compile
# To update, run:
#
#    PYTHON_VERSION=3 EXTRAS=coverage,orchestra,test generate-requirements.sh
#
```

toabctl replied:
you would need a requirements file for every supported python3 version because the output of `pip-compile` might be different for different python3 minor versions...

djgalloway commented:
This PR broke the workers in the lab.

```
2019-12-06T16:20:02.532 INFO:teuthology.repo_utils:Cloning git://git.ceph.com/git/teuthology.git master from upstream
2019-12-06T16:20:02.554 CRITICAL:teuthology:Uncaught exception
Traceback (most recent call last):
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/virtualenv/bin/teuthology-worker", line 11, in <module>
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/scripts/worker.py", line 7, in main
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/worker.py", line 120, in main
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/worker.py", line 156, in prep_job
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/repo_utils.py", line 387, in fetch_teuthology
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/repo_utils.py", line 324, in fetch_repo
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/repo_utils.py", line 87, in enforce_repo_state
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/repo_utils.py", line 127, in clone_repo
  File "/usr/lib/python2.7/subprocess.py", line 711, in __init__
    errread, errwrite)
  File "/usr/lib/python2.7/subprocess.py", line 1235, in _execute_child
    self.pid = os.fork()
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/virtualenv/local/lib/python2.7/site-packages/gevent/os.py", line 445, in fork
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/virtualenv/local/lib/python2.7/site-packages/gevent/os.py", line 174, in fork_gevent
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/virtualenv/local/lib/python2.7/site-packages/gevent/hub.py", line 334, in reinit
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/virtualenv/local/lib/python2.7/site-packages/gevent/hub.py", line 767, in _get_threadpool
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/virtualenv/local/lib/python2.7/site-packages/gevent/hub.py", line 399, in _import
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/virtualenv/local/lib/python2.7/site-packages/gevent/hub.py", line 415, in _import
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/virtualenv/local/lib/python2.7/site-packages/gevent/builtins.py", line 93, in __import__
ImportError: No module named threadpool
2019-12-06T16:20:03.164 ERROR:teuthology.repo_utils:Git error encountered; retrying
Traceback (most recent call last):
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/repo_utils.py", line 324, in fetch_repo
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/repo_utils.py", line 87, in enforce_repo_state
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/repo_utils.py", line 142, in clone_repo
GitError: git clone failed!
```

kshtsk commented:
@djgalloway does this logs comes from the worker or from the job itself?

djgalloway commented:
That's a snippet from the worker log.  It just repeats that for 10 minutes until the job gets marked dead and another worker does the same thing.

```
2019-12-06T16:29:57.819 ERROR:teuthology.worker:Failed to fetch or bootstrap; marking job as dead
Traceback (most recent call last):
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/worker.py", line 156, in prep_job
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/repo_utils.py", line 387, in fetch_teuthology
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/repo_utils.py", line 322, in fetch_repo
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/contextutil.py", line 134, in __call__
MaxWhileTries: reached maximum tries (60) after waiting for 600 seconds
```

kshtsk commented:
I haven't seen this issue on the deployed teuthology from scratch, probably this issue can go away after worker restarts on the new virtualenv, is it possible to try and restart all the workers?

djgalloway commented:
I `touch`ed `/tmp/teuthology-restart-workers` 20min ago which should have allowed at least one worker to restart and re-bootstrap but workers are still hung.

djgalloway commented:
I just killed all the PIDs waiting on `/home/teuthworker/src/git.ceph.com_git_teuthology_master.lock`, started a new workers, and it appears to be running a job.

tchaikov commented:
thanks @kshtsk and @djgalloway !
